GitUI/UserControls/WarningHeader.cs

```diff
@@ -0,0 +1,193 @@
+﻿using System;
+using System.Threading.Tasks;
+using System.Windows.Forms;
+using GitCommands;
+using GitUI.CommandsDialogs;
+using Microsoft.VisualStudio.Threading;
+
+namespace GitUI.UserControls
+{
+    public partial class WarningHeader : UserControl
+    {
+        private enum State
+        {
+            None,
+            Bisect,
+            Rebase,
+            Merge,
+            Patch
+        }
+
+        private State _state = State.None;
+        private bool _hasConflicts = false;
+
+        public event EventHandler BisectEvent;
+
+        public WarningHeader()
+        {
+            InitializeComponent();
+        }
+
+        public async Task RefreshAsync(GitCommands.GitModule module)
+        {
+            // get the current state of the repo
+            await TaskScheduler.Default;
+
+            if (module.IsValidGitWorkingDir())
+            {
+                if (module.InTheMiddleOfBisect())
+                {
+                    await SetStateAsync(State.Bisect, false);
```

Comment:
Add `return` here for fast return and remove unnecessary `else`
Applies everywhere.